### PR TITLE
add error logging and metrics

### PR DESCRIPTION
**Pull request by asusac**
https://trello.com/c/fsbfHae0/3349-add-error-logging-and-metrics-to-turnilo

[add error logging and metrics to turnilo](https://trello.com/c/fsbfHae0/3349-add-error-logging-and-metrics-to-turnilo)

### CODE REVIEW

File: src/server/app.ts [hunk 1]
```diff
@@ -48,6 +51,27 @@ function addRoutes(attach: string, router: Router | Handler): void {
   app.use(SERVER_SETTINGS.getServerRoot() + attach, router);
 }
 
+// Add request metrics
+app.use((req: Request, res: Response, done) => {
+  const stopRequestTimer = metrics.requestDuration.startTimer();
+
+  function addRequestMetrics() {
+    const lables = { host: req.hostname, route: req.path, method: req.method, status: res.statusCode };
```
Comment (itszootime) [inline src/server/app.ts]: The `host` label won't be very useful as the proxy does not rewrite the `Host` header. We can simply remove this as Prometheus adds the `instance` label automatically. My fault for not thinking of this before, sorry.

File: src/server/app.ts [hunk 2]
```diff
@@ -48,6 +51,27 @@ function addRoutes(attach: string, router: Router | Handler): void {
   app.use(SERVER_SETTINGS.getServerRoot() + attach, router);
 }
 
+// Add request metrics
```
Comment (itszootime) [inline src/server/app.ts]: As this also adds metrics for static files, maybe we can limit it to the `/plywood` endpoint?

File: src/server/app.ts [hunk 3]
```diff
@@ -48,6 +51,27 @@ function addRoutes(attach: string, router: Router | Handler): void {
...
+  function addRequestMetrics() {
+    const lables = { host: req.hostname, route: req.path, method: req.method, status: res.statusCode };
+    res.removeListener("finish", addRequestMetrics);
+    res.removeListener("close", addRequestMetrics);
+
+    metrics.totalRequests.inc(lables);
+    stopRequestTimer(lables);
+  }
+
+  res.on("finish", addRequestMetrics);
```
Comment (itszootime) [inline src/server/app.ts]: It seems using [`on-finished`](https://github.com/jshttp/on-finished) is now the recommended way to do this. We then shouldn't need to use any other `on` or `removeListener` functions.

**Comment by asusac** [comment]
used different implementation to diverge less from upstream

**Pull request closed without merging**